components/script/dom/textdecodercommon.rs

```diff
@@ -0,0 +1,217 @@
+        self.encoding
+    }
+
+    /// <https://encoding.spec.whatwg.org/#textdecodercommon-decoder>
+    pub(crate) fn decoder(&self) -> &RefCell<Decoder> {
+        &self.decoder
+    }
+
+    /// <https://encoding.spec.whatwg.org/#textdecodercommon-i-o-queue>
+    pub(crate) fn io_queue(&self) -> &RefCell<Vec<u8>> {
+        &self.io_queue
+    }
+
+    /// <https://encoding.spec.whatwg.org/#textdecoder-error-mode>
+    pub(crate) fn fatal(&self) -> bool {
+        self.fatal
+    }
+
+    /// <https://encoding.spec.whatwg.org/#textdecoder-ignore-bom-flag>
+    pub(crate) fn ignore_bom(&self) -> bool {
+        self.ignoreBOM
+    }
+
+    /// Shared by `TextDecoder` and `TextDecoderStream`
+    ///
+    /// <https://encoding.spec.whatwg.org/#dom-textdecoder-decode>
+    /// <https://encoding.spec.whatwg.org/#decode-and-enqueue-a-chunk>
+    #[allow(unsafe_code)]
+    pub(crate) fn decode(
+        &self,
+        input: Option<&ArrayBufferViewOrArrayBuffer>,
+        last: bool,
+    ) -> Fallible<String> {
+        // <https://encoding.spec.whatwg.org/#dom-textdecoder-decode>
+        // Step 3. If input is given, then push a copy of input to this’s I/O queue.
+        //
+        // <https://encoding.spec.whatwg.org/#decode-and-enqueue-a-chunk>
+        // Step 2. Push a copy of bufferSource to decoder’s I/O queue.
+        //
+        // NOTE: try to avoid this copy unless there are bytes left
+        let mut io_queue = self.io_queue.borrow_mut();
+        let input = match &input {
+            Some(ArrayBufferViewOrArrayBuffer::ArrayBufferView(a)) => unsafe {
+                if io_queue.is_empty() {
+                    a.as_slice()
+                } else {
+                    io_queue.extend_from_slice(a.as_slice());
+                    &io_queue[..]
+                }
+            },
+            Some(ArrayBufferViewOrArrayBuffer::ArrayBuffer(a)) => unsafe {
+                if io_queue.is_empty() {
+                    a.as_slice()
+                } else {
+                    io_queue.extend_from_slice(a.as_slice());
+                    &io_queue[..]
+                }
+            },
+            None => &io_queue[..],
+        };
+
+        let mut decoder = self.decoder.borrow_mut();
+        let (output, read) = if self.fatal {
+            // <https://encoding.spec.whatwg.org/#dom-textdecoder-decode>
+            // Step 4. Let output be the I/O queue of scalar values « end-of-queue ».
+            //
+            // <https://encoding.spec.whatwg.org/#decode-and-enqueue-a-chunk>
+            // Step 3. Let output be the I/O queue of scalar values « end-of-queue ».
+            let mut output = String::with_capacity(
+                decoder
+                    .max_utf8_buffer_length_without_replacement(input.len())
+                    .expect("Expected UTF8 buffer length would overflow"),
+            );
+
+            // <https://encoding.spec.whatwg.org/#dom-textdecoder-decode>
```

Comment:
let's add a note that `decode_to_string_without_replacement` implements the steps of the two below algorithms.